File: ext/gantifontshighlight.pl
#!/usr/bin/env perl
use strict;
use warnings;
use File::Copy qw(copy);
use Getopt::Long;

# ===============================
# 🧩 Opsi CLI
# ===============================
my $quiet = 0;
my $no_backup = 0;

GetOptions(
  "quiet"      => \$quiet,
  "no-backup"  => \$no_backup,
);

# ===============================
# 🔁 Peta penggantian CDN → lokal
# ===============================
my %REPLACEMENTS = (
  # --- Font Awesome (semua CDN, semua versi)
  qr{href=["']https://(?:cdnjs\.cloudflare\.com/ajax/libs|cdn\.jsdelivr\.net/npm/\@fortawesome/fontawesome-free\@[^/]+|use\.fontawesome\.com/releases/v[\d\.]+)/css/all\.min\.css["'][^>]*}i
    => 'href="/ext/fontawesome.css"',

  # --- Highlight.js JS (semua sumber)
  qr{src=["']https://(?:cdnjs\.cloudflare\.com/ajax/libs/highlight\.js|cdn\.jsdelivr\.net/(?:npm/highlight\.js\@[^/]+|gh/highlightjs/cdn-release\@[^/]+/build))/highlight\.min\.js["'][^>]*}i
    => 'src="/ext/highlight.js"',

  # --- Highlight.js CSS (tema terang)
  qr{href=["']https://(?:cdnjs\.cloudflare\.com/ajax/libs/highlight\.js|cdn\.jsdelivr\.net/(?:npm/highlight\.js\@[^/]+|gh/highlightjs/cdn-release\@[^/]+/build))/styles/github\.min\.css["'][^>]*}i
    => 'href="/ext/github.min.css"',

  # --- Highlight.js CSS (tema gelap)
  qr{href=["']https://(?:cdnjs\.cloudflare\.com/ajax/libs/highlight\.js|cdn\.jsdelivr\.net/(?:npm/highlight\.js\@[^/]+|gh/highlightjs/cdn-release\@[^/]+/build))/styles/github-dark\.min\.css["'][^>]*}i
    => 'href="/ext/github-dark.min.css"',

  # --- Highlight.js CSS default (misalnya default.min.css)
  qr{href=["']https://(?:cdnjs\.cloudflare\.com/ajax/libs/highlight\.js|cdn\.jsdelivr\.net/(?:npm/highlight\.js\@[^/]+|gh/highlightjs/cdn-release\@[^/]+/build))/styles/default\.min\.css["'][^>]*}i
    => 'href="/ext/github-dark.min.css"',
);

# ===============================
# 📁 Cari file HTML
# ===============================
my @files = glob("*.html artikelx/*.html artikel/*.html");

if (!@files) {
  print "⚠️  Tidak ada file HTML ditemukan.\n" unless $quiet;
  exit 0;
}

foreach my $file (@files) {
  local $/ = undef;
  open my $fh, "<:utf8", $file or next;
  my $content = <$fh>;
  close $fh;

  my $changed = 0;

  while (my ($regex, $replacement) = each %REPLACEMENTS) {
    my $count = ($content =~ s/$regex/$replacement/gi);
    $changed += $count if $count;
  }

  if ($changed > 0) {
    unless ($no_backup) {
      copy($file, "$file.bak");
      print "🗂️  Backup dibuat: $file.bak\n" unless $quiet;
    }

    open my $out, ">:utf8", $file or next;
    print $out $content;
    close $out;

    print "✅ $file — $changed tautan CDN diganti ke /ext lokal\n" unless $quiet;
  } else {
    print "⏭️  Tidak ada penggantian di $file\n" unless $quiet;
  }
}

print "🎯 Selesai. Semua CDN kini diarahkan ke aset lokal.\n" unless $quiet;

File: ext/gantifontshighlight.js
import { Glob } from "bun";

// Konfigurasi CLI
const args = Bun.argv.slice(2);
const NO_BACKUP = args.includes("--no-backup");

// Mapping yang persis dengan daftar Perl kamu
const MAP = [
    { rx: /https:\/\/.*?\/(?:all|fontawesome)(\.min)?\.css/i, repl: '/ext/fontawesome.css' },
    { rx: /https:\/\/.*?\/(?:prism|vs-dark|default|github|monokai|atom-one).*?\.css/i, repl: '/ext/default.min.css' }, // Catch-all CSS
    { rx: /https:\/\/.*?\/leaflet\.css/i, repl: '/ext/leaflet.css' },
    { rx: /https:\/\/.*?\/highlight\.min\.js/i, repl: '/ext/highlight.js' }
];

// Regex Sapu Jagat untuk menangkap atribut href/src DAN membersihkan sisa sampah (integrity, dsb)
// Ini adalah versi JS dari regex 'while' milik Perl kamu
const SUPER_REGEX = /(\b(?:href|src)\b)\s*=\s*(['"])\s*https?:\/\/[^"']+?\/([^"']+?\.(?:css|js))(?:\?[^"']*)?\s*\2([^>]*)/gi;

async function processFile(filePath) {
    if (filePath.endsWith("index.html")) return;
    
    const file = Bun.file(filePath);
    let content = await file.text();
    let changed = false;

    // 1. Eksekusi Penggantian URL & Pembersihan Atribut sekaligus (Mirip Perl)
    content = content.replace(SUPER_REGEX, (match, attr, quote, fileName, extraAttr) => {
        let replacement = null;

        // Cek apakah file ini ada di MAP
        for (const m of MAP) {
            if (match.match(m.rx)) {
                replacement = m.repl;
                break;
            }
        }

        if (replacement) {
            changed = true;
            // Kita kembalikan hanya atribut utama, 'extraAttr' (integrity, crossorigin) dibuang!
            return `${attr}=${quote}${replacement}${quote}`;
        }
        
        return match; // Tidak ada yang cocok, kembalikan aslinya
    });

    if (changed) {
        if (!NO_BACKUP) await Bun.write(`${filePath}.bak`, await file.text());
        await Bun.write(filePath, content);
        console.log(`✅ Fixed: ${filePath}`);
    }
}

async function run() {
    console.log("🚀 Menantang Perl: Memulai pemindaian...");
    const glob = new Glob("{*.html,artikelx/*.html,artikel/*.html}");
    
    const files = [];
    for await (const file of glob.scan(".")) files.push(file);
    
    await Promise.all(files.map(processFile));
    console.log("✨ Selesai! Link cdnjs font-awesome 6.4.0? Tumbang!");
}

run();
